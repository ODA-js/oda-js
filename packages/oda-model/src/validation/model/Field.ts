import { Map } from 'immutable';
import { Record } from 'immutable';

import {
  IsBelongsTo,
  IsBelongsToMany,
  IsBelongsToManyProps,
  IsBelongsToProps,
  IsHasMany,
  IsHasManyProps,
  IsHasOne,
  IsHasOneProps,
} from '../helpers';
import { IBelongsToInit } from '../interfaces/IBelongsTo';
import { IBelongsToManyInit } from '../interfaces/IBelongsToMany';
import { IEntityRef } from '../interfaces/IEntityRef';
import { IField } from '../interfaces/IField';
import { IFieldACL, IFieldArgs, IFieldInit, IFieldStore, IFieldTransform } from '../interfaces/IField';
import { IHasManyInit } from '../interfaces/IHasMany';
import { IHasOneInit } from '../interfaces/IHasOne';
import { IRelation, IRelationInit } from '../interfaces/IRelation';
import { BelongsTo, BelongsToTransform } from './BelongsTo';
import { BelongsToMany, BelongsToManyTransform } from './BelongsToMany';
import { HasMany, HasManyTransform } from './HasMany';
import { HasOne, HasOneTransform } from './HasOne';
import { Persistent } from './Persistent';
import { transformMap } from './utils';
import { IEntity } from '../interfaces/IEntity';

// tslint:disable-next-line:variable-name
export const DefaultField: IFieldStore = {
  name: null,
  title: null,
  description: null,
  acl: null,
  args: null,
  derived: null,
  entity: null,
  identity: null,
  idKey: null,
  indexed: null,
  order: null,
  persistent: null,
  relation: null,
  required: null,
  type: null,
};

// tslint:disable-next-line:variable-name
export const FieldTransform: IFieldTransform = {
  args: transformMap<IFieldArgs>(),
  relation: {
    transform: (inp: IRelationInit): IRelation => {
      if (IsBelongsToProps(inp)) {
        return new BelongsTo(inp);
      }
      if (IsBelongsToManyProps(inp)) {
        return new BelongsToMany(inp);
      }
      if (IsHasOneProps(inp)) {
        return new HasOne(inp);
      }
      if (IsHasManyProps(inp)) {
        return new HasMany(inp);
      }
    },
    reverse: (inp: IRelation): IRelationInit => {
      if (IsBelongsTo(inp)) {
        return {
          ...inp,
          belongsTo: BelongsToTransform.belongsTo.reverse(inp.belongsTo),
          fields: BelongsToTransform.fields.reverse(inp.fields),
        } as IBelongsToInit;
      }
      if (IsBelongsToMany(inp)) {
        return {
          ...inp,
          fields: BelongsToManyTransform.fields.reverse(inp.fields),
        } as IBelongsToManyInit;
      }
      if (IsHasOne(inp)) {
        return {
          ...inp,
          fields: HasOneTransform.fields.reverse(inp.fields),
        } as IHasOneInit;
      }
      if (IsHasMany(inp)) {
        return {
          ...inp,
          fields: HasManyTransform.fields.reverse(inp.fields),
        } as IHasManyInit;
      }
    },
  },
};

// tslint:disable-next-line:variable-name
export const FieldStorage = Record(DefaultField);

export class Field extends Persistent<IFieldInit, IFieldStore> implements IField {
  public get modelType(): 'field' {
    return 'field';
  }
  public get entity(): IEntity {
    return this.store.get('entity', null);
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get type(): string {
    return this.store.get('type', null);
  }
  public get acl(): Partial<IFieldACL> {
    return this.store.get('acl', null);
  }
  public get args(): Map<string, IFieldArgs> {
    return this.store.get('args', null);
  }
  public get derived(): boolean {
    return this.store.get('derived', null);
  }
  public get persistent(): boolean {
    return this.store.get('persistent', null);
  }
  public get required(): boolean {
    return this.store.get('required', null);
  }
  public get indexed(): boolean | string | string[] {
    return this.store.get('indexed', null);
  }
  public get identity(): boolean | string | string[] {
    return this.store.get('identity', null);
  }
  public get idKey(): IEntityRef {
    return this.store.get('idKey', null);
  }
  public get order(): number {
    return this.store.get('order', null);
  }
  public get relation(): IRelation {
    return this.store.get('relation', null);
  }
  protected transform(input: IFieldInit): IFieldStore {
    const result: IFieldStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'args') {
            result.args = FieldTransform.args.transform(input.args);
          } else if (f === 'relation') {
            result.relation = FieldTransform.relation.transform(input.relation);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }
  protected reverse(input: IFieldStore): IFieldInit {
    const result: IFieldInit = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'args') {
            result.args = FieldTransform.args.reverse(input.args);
          } else if (f === 'relation') {
            result.relation = FieldTransform.relation.reverse(input.relation);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }
  constructor(init: IFieldInit) {
    super();
    this.store = new FieldStorage(this.transform(init));
    this.init = new (Record<IFieldInit>(init))();
  }
}
