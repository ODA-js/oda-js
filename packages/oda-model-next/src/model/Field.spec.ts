import 'jest';
import { Field } from './Field';

describe('Field', () => {
  let field: Field;

  beforeAll(() => {
    expect(() => new Field()).not.toThrow();
  });

  beforeEach(() => {
    field = new Field();
  });

  it('create empty', () => {
    expect(field).toMatchSnapshot();
    expect(field.toJS()).toMatchSnapshot();
  });

  it('update strings', () => {
    expect(() =>
      field.updateWith({
        name: 'cool',
        description: 'very cool',
        title: 'very cool title'
      })
    ).not.toThrow();
    expect(field.toJS()).toMatchSnapshot();
  });

  it('updates with null or undefined', () => {
    expect(() =>
      field.updateWith({
        name: 'cool',
        description: 'very cool',
        title: 'very cool title'
      })
    ).not.toThrow();
    expect(field.toJS()).toMatchSnapshot();
    expect(() =>
      field.updateWith({
        description: null,
        title: undefined
      })
    ).not.toThrow();
    expect(field.toJS()).toMatchSnapshot();
  });

  it('toJS with dupes', () => {
    expect(() =>
      field.updateWith({
        name: 'cool',
        description: 'very cool',
        title: 'very cool title',
        args: [{ name: 'one' }, { name: 'one' }]
      })
    ).not.toThrow();

    expect(field.args.size).toBe(1);
    expect(field.toJS()).toMatchSnapshot();
  });

  it('toJS with relations hasMany', () => {
    expect(() =>
      field.updateWith({
        name: 'cool',
        description: 'very cool',
        title: 'very cool title',
        relation: {
          hasMany: 'id@some#id'
        }
      })
    ).not.toThrow();

    expect(field.relation).not.toBeNull();
    expect(field.toJS()).toMatchSnapshot();
  });

  it('toJS with relations hasOne', () => {
    expect(() =>
      field.updateWith({
        name: 'cool',
        description: 'very cool',
        title: 'very cool title',
        relation: {
          hasOne: 'id@some#id'
        }
      })
    ).not.toThrow();

    expect(field.relation).not.toBeNull();
    expect(field.toJS()).toMatchSnapshot();
  });

  it('toJS with relations belongsTo', () => {
    expect(() =>
      field.updateWith({
        name: 'cool',
        description: 'very cool',
        title: 'very cool title',
        relation: {
          belongsTo: 'id@some#id'
        }
      })
    ).not.toThrow();

    expect(field.relation).not.toBeNull();
    expect(field.toJS()).toMatchSnapshot();
  });

  it('toJS with relations belongsToMany', () => {
    expect(() =>
      field.updateWith({
        name: 'cool',
        description: 'very cool',
        title: 'very cool title',
        relation: {
          belongsToMany: 'id@some#id',
          using: 'Some@at#this'
        }
      })
    ).not.toThrow();

    expect(field.relation).not.toBeNull();
    expect(field.toJS()).toMatchSnapshot();
  });
});
