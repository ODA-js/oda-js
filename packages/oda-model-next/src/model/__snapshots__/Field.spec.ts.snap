// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field create empty 1`] = `
Field {
  "store": Immutable.Record {
    "name": null,
    "title": null,
    "description": null,
    "acl": null,
    "args": null,
    "derived": null,
    "entity": null,
    "identity": null,
    "idKey": null,
    "indexed": null,
    "order": null,
    "persistent": null,
    "relation": null,
    "required": null,
    "type": null,
  },
}
`;

exports[`Field create empty 2`] = `Object {}`;

exports[`Field toJS with dupes 1`] = `
Object {
  "args": Array [
    Object {
      "name": "one",
    },
  ],
  "description": "very cool",
  "name": "cool",
  "title": "very cool title",
}
`;

exports[`Field toJS with relations belongsTo 1`] = `
Object {
  "description": "very cool",
  "name": "cool",
  "relation": Object {
    "belongsTo": "id@Some#id",
    "fields": null,
  },
  "title": "very cool title",
}
`;

exports[`Field toJS with relations belongsToMany 1`] = `
Object {
  "description": "very cool",
  "name": "cool",
  "relation": Object {
    "belongsToMany": "id@Some#id",
    "fields": null,
    "using": "Some@At#this",
  },
  "title": "very cool title",
}
`;

exports[`Field toJS with relations hasMany 1`] = `
Object {
  "description": "very cool",
  "name": "cool",
  "relation": Object {
    "fields": null,
    "hasMany": "id@Some#id",
  },
  "title": "very cool title",
}
`;

exports[`Field toJS with relations hasOne 1`] = `
Object {
  "description": "very cool",
  "name": "cool",
  "relation": Object {
    "fields": null,
    "hasOne": "id@Some#id",
  },
  "title": "very cool title",
}
`;

exports[`Field update strings 1`] = `
Object {
  "description": "very cool",
  "name": "cool",
  "title": "very cool title",
}
`;

exports[`Field updates with null or undefined 1`] = `
Object {
  "description": "very cool",
  "name": "cool",
  "title": "very cool title",
}
`;

exports[`Field updates with null or undefined 2`] = `
Object {
  "name": "cool",
}
`;
